OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ * Author(s): Ajay Seth                                                       *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** @file
+* This file defines the  DataTable class, which is used by OpenSim to provide
+* an in memory container for data access and manipulation.
+*/
+#include "SimTKmath.h"
+#include "XMLDocument.h"
+#include "FileAdapter.h"
+#include "Array.h"
+#include <memory>
+#include <typeindex>
+
+
+namespace OpenSim {
+
+//=============================================================================
+/**
+ * AbstractDataTable defines a simple interface for in memory access to
+ * numerical data as a table. A DataTable is independent of the data source 
+ * used to populate it. A concrete DataTable provides fast access to data
+ * elements by row and/or column indices as well as columns by label.
+ *
+ * @author Ajay Seth
+ */
+class AbstractDataTable {
+public:
+    virtual ~AbstractDataTable() = default;
+
+    virtual AbstractDataTable* clone() const = 0;
+
+    /** Read the meta data associated with this DataTable as an XMLDocument */
+    virtual const XMLDocument& getMetaData() const = 0;
+    /** Update the meta data associated with this DataTable as an XMLDocument */
+    virtual XMLDocument& updMetaData() = 0;
+
+    /** Get the number of rows in the DataTable */
+    virtual size_t  getNumRows() const = 0;
+    /** Get the number of columns in the DataTable */
+    virtual size_t  getNumCols() const = 0;
+
+    /** Determine if the DataTable has a column associated with a 
+        column label (string) */
+    virtual bool hasColumn(const std::string& label) const = 0;
+    /** For a known column label return its column index. */
+    virtual size_t getColumnIndex(const std::string& label) const = 0;
```

Comment:
* Can we name hasColumn to hasCol ?
* Can we name getColumnIndex to colInd ? Here again, there is going to be no "set" counterpart.